juju/juju
# [JUJU-2420] Use model-config instead of environment variable to control prompting globally

We decided that environment variables were a bad idea for this feature. They're a bit magic and also potentially dangerous

Since the requirement for this extra layer of protection is is property of models, add this setting as a value for the (currently unused) 'mode' config option

By default for 3.0 and 3.1 this value will be absent. However, this will change in 3.2 

The first three commits are back-ported from develop, where we abstracted out the prompting behaviour. However, this abstraction is already used by destroy-* commands which (for 3.0 and 3.1 only) behave differently. So there is still some (temporary) duplication here.

## Checklist


## QA steps

Verify the config option `mode` is present and defaults to the empty string

for each of `remove-machine`, `remove-application` and `remove-unit` verify that:
- When `mode` doesn't contains `requires-prompts`, no prompting occurs
- When `mode` does contains `requires-prompts`, prompting occurs
  - This prompt can be removed using the `--no-prompt` flag

Verify nothing breaks for a 3.0.2 model

Verify that destroy-* commands still function as before
